File: postman/enhance_collection.py
#!/usr/bin/env python3
"""
EchoVault Postman Collection Enhancer
Adds test scripts, examples, and validation to all request files
"""

import os
import yaml
from pathlib import Path

# Define test scripts for different request types
TEST_SCRIPTS = {
    "Auth": {
        "Register": """pm.test("Status code is 200 or 201", function () {
  pm.response.to.have.status([200, 201]);
});

pm.test("Response has token", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("token");
  pm.expect(json.token).to.be.a("string");
  if (json.token) pm.environment.set("token", json.token);
});

pm.test("Response has user data", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("user");
  pm.expect(json.user).to.have.property("id");
  pm.expect(json.user).to.have.property("email");
});""",
        
        "Login": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response has token", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("token");
  pm.expect(json.token).to.be.a("string");
  pm.environment.set("token", json.token);
  pm.environment.set("artistToken", json.token);
});

pm.test("Response has user details", function () {
  const json = pm.response.json();
  pm.expect(json.user).to.have.property("id");
  pm.expect(json.user).to.have.property("email");
  pm.expect(json.user).to.have.property("role");
});""",
        
        "Logout": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response confirms logout", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("success");
  pm.expect(json.success).to.be.true;
});""",
        
        "Refresh Token": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("New token in response", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("token");
  pm.expect(json.token).to.be.a("string");
  pm.environment.set("token", json.token);
});""",
        
        "Verify Auth": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("User is verified", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("user");
  pm.expect(json.user).to.have.property("id");
});""",
    },
    
    "Artist": {
        "GET": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response is valid JSON", function () {
  pm.response.to.have.header("content-type");
  pm.expect(pm.response.json()).to.be.an("object");
});

pm.test("Has required data fields", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("data");
});""",
        
        "POST": """pm.test("Status code is 201 or 200", function () {
  pm.response.to.have.status([200, 201]);
});

pm.test("Response has success message", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("success");
  pm.expect(json.success).to.be.true;
});""",
    },
    
    "Admin": {
        "GET": """pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response is valid", function () {
  pm.expect(pm.response.json()).to.be.an("object");
});

pm.test("Admin authorization enforced", function () {
  // If returns 403, admin auth is properly enforced
  pm.expect([200, 403]).to.include(pm.response.code);
});""",
        
        "POST": """pm.test("Status code is 200 or 201", function () {
  pm.response.to.have.status([200, 201]);
});

pm.test("Operation successful", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("success");
  pm.expect(json.success).to.be.true;
});""",
    },
}

# Example responses by endpoint type
EXAMPLE_RESPONSES = {
    "Login": {
        "status": 200,
        "body": {
            "success": True,
            "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
            "user": {
                "id": "user123",
                "email": "artist@example.com",
                "name": "John Artist",
                "role": "artist",
                "verified": True
            }
        }
    },
    "Register": {
        "status": 201,
        "body": {
            "success": True,
            "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...",
            "user": {
                "id": "user123",
                "email": "newartist@example.com",
                "name": "New Artist",
                "role": "artist"
            }
        }
    },
    "Get Insights": {
        "status": 200,
        "body": {
            "success": True,
            "data": {
                "totalStreams": 15000,
                "totalRevenue": 450.50,
                "monthlyGrowth": 12.5,
                "topTrack": {
                    "id": "track1",
                    "title": "My Hit Song",
                    "streams": 5000
                }
            }
        }
    },
    "Get Dashboard": {
        "status": 200,
        "body": {
            "success": True,
            "data": {
                "stats": {
                    "totalStreams": 25000,
                    "totalRevenue": 750.75,
                    "activeListeners": 350
                },
                "recentActivity": []
            }
        }
    },
    "Get All Users": {
        "status": 200,
        "body": {
            "success": True,
            "data": {
                "users": [
                    {
                        "id": "user1",
                        "email": "artist1@example.com",
                        "name": "Artist One",
                        "role": "artist",
                        "verified": True,
                        "createdAt": "2024-01-15T10:30:00Z"
                    }
                ],
                "total": 1,
                "page": 1,
                "limit": 10
            }
        }
    },
    "Verify Artist": {
        "status": 200,
        "body": {
            "success": True,
            "message": "Artist verified successfully",
            "artist": {
                "id": "artist123",
                "email": "artist@example.com",
                "verified": True
            }
        }
    }
}

def get_test_script(folder, filename):
    """Get appropriate test script for endpoint"""
    if folder == "Auth":
        name = filename.replace(".request.yaml", "")
        return TEST_SCRIPTS["Auth"].get(name, TEST_SCRIPTS["Artist"]["GET"])
    elif folder in ["Artist", "Artist Dashboard"]:
        if "Get" in filename or "Insights" in filename or "Library" in filename:
            return TEST_SCRIPTS["Artist"]["GET"]
        else:
            return TEST_SCRIPTS["Artist"]["POST"]
    elif folder == "Admin":
        if "Get" in filename or "Unverified" in filename:
            return TEST_SCRIPTS["Admin"]["GET"]
        else:
            return TEST_SCRIPTS["Admin"]["POST"]
    elif folder == "Tracks":
        return TEST_SCRIPTS["Artist"]["GET"]
    elif folder == "Analytics":
        if "Get" in filename:
            return TEST_SCRIPTS["Artist"]["GET"]
        else:
            return TEST_SCRIPTS["Artist"]["POST"]
    else:
        return TEST_SCRIPTS["Artist"]["GET"]

def get_example_response(filename):
    """Get example response for endpoint"""
    for key, response in EXAMPLE_RESPONSES.items():
        if key in filename:
            return response
    
    # Default example
    return {
        "status": 200,
        "body": {
            "success": True,
            "data": {}
        }
    }

def enhance_request_file(filepath):
    """Add test scripts and examples to request file"""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = yaml.safe_load(f)
        
        # Add test script
        folder = filepath.parent.parent.name
        filename = filepath.name
        
        test_code = get_test_script(folder, filename)
        
        if "scripts" not in content:
            content["scripts"] = []
        
        # Remove existing test scripts
        content["scripts"] = [s for s in content.get("scripts", []) if s.get("type") != "afterResponse"]
        
        # Add new test script
        content["scripts"].append({
            "type": "afterResponse",
            "language": "text/javascript",
            "code": test_code
        })
        
        # Add example response
        example = get_example_response(filename)
        if "examples" not in content:
            content["examples"] = []
        
        content["examples"] = [{
            "name": f"{filename.replace('.request.yaml', '')} - Success",
            "status": example["status"],
            "headers": [
                {"key": "Content-Type", "value": "application/json"}
            ],
            "body": example["body"]
        }]
        
        # Write back
        with open(filepath, 'w', encoding='utf-8') as f:
            yaml.dump(content, f, default_flow_style=False, sort_keys=False)
        
        return True, "Enhanced"
    except Exception as e:
        return False, str(e)

def main():
    base_path = Path("postman/collections/EchoVault API")
    
    if not base_path.exists():
        print(f"Collection path not found: {base_path}")
        return
    
    # Find all .request.yaml files
    request_files = list(base_path.rglob("*.request.yaml"))
    
    print(f"Found {len(request_files)} request files\n")
    
    success_count = 0
    for filepath in sorted(request_files):
        folder = filepath.parent.name
        filename = filepath.name
        
        success, msg = enhance_request_file(filepath)
        
        status = "✓" if success else "✗"
        print(f"{status} {folder:20} | {filename:40} | {msg}")
        
        if success:
            success_count += 1
    
    print(f"\n{'='*80}")
    print(f"✓ Enhanced {success_count}/{len(request_files)} request files")
    print(f"✓ Added test scripts to all endpoints")
    print(f"✓ Added example responses to all endpoints")

if __name__ == "__main__":
    main()

File: postman/enhance_collection.js
#!/usr/bin/env node
/**
 * EchoVault Postman Collection Enhancer
 * Adds test scripts, examples, and validation to all request files
 * Uses pure Node.js (no external dependencies)
 */

const fs = require('fs');
const path = require('path');

// Simple YAML parser (very basic, for our use case)
function parseYAML(content) {
  const lines = content.split('\n');
  const result = {};
  let current = result;
  let stack = [{ obj: result, indent: 0 }];
  let inMultiline = false;
  let multilineKey = '';
  let multilineIndent = 0;
  let multilineContent = [];

  for (let i = 0; i < lines.length; i++) {
    let line = lines[i];
    
    // Handle multiline content (preserve as-is)
    if (inMultiline) {
      const indent = line.search(/\S/);
      if (line.trim() === '') {
        multilineContent.push('');
      } else if (indent > multilineIndent || line.trim().startsWith('- ') || line.trim().startsWith('{')) {
        multilineContent.push(line.slice(multilineIndent));
      } else {
        // End of multiline
        inMultiline = false;
        current[multilineKey] = multilineContent.join('\n');
        i--;
        continue;
      }
    } else if (line.trim().endsWith('|-')) {
      // Start of multiline literal
      const [key] = line.split(':');
      multilineKey = key.trim();
      multilineIndent = line.search(/\S/) + 2;
      multilineContent = [];
      inMultiline = true;
      continue;
    }

    if (!inMultiline) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;

      const indent = line.search(/\S/);
      
      // Adjust stack based on indent
      while (stack.length > 1 && stack[stack.length - 1].indent >= indent) {
        stack.pop();
      }
      current = stack[stack.length - 1].obj;

      if (trimmed.includes(':')) {
        const [key, ...valueParts] = trimmed.split(':');
        const value = valueParts.join(':').trim();
        
        if (!value || value === '[]' || value === '{}') {
          // Complex value
          const newObj = value === '{}' ? {} : [];
          current[key.trim()] = newObj;
          stack.push({ obj: newObj, indent });
        } else if (value.startsWith('[') && value.endsWith(']')) {
          current[key.trim()] = JSON.parse(value);
        } else if (value === 'true' || value === 'false') {
          current[key.trim()] = value === 'true';
        } else if (!isNaN(value)) {
          current[key.trim()] = parseFloat(value);
        } else if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
          current[key.trim()] = value.slice(1, -1);
        } else {
          current[key.trim()] = value;
        }
      } else if (trimmed.startsWith('- ')) {
        if (Array.isArray(current)) {
          const val = trimmed.slice(2).trim();
          current.push(val);
        }
      }
    }
  }

  if (inMultiline) {
    current[multilineKey] = multilineContent.join('\n');
  }

  return result;
}

// Simple YAML stringifier
function stringifyYAML(obj, indent = 0) {
  const spaces = '  '.repeat(indent);
  let result = '';

  for (const [key, value] of Object.entries(obj)) {
    if (value === null || value === undefined) continue;

    if (typeof value === 'string') {
      if (value.includes('\n')) {
        result += `${spaces}${key}: |-\n`;
        result += value.split('\n').map(line => `${spaces}  ${line}`).join('\n') + '\n';
      } else if (value.includes(':') || value.includes('{') || value.startsWith("'")) {
        result += `${spaces}${key}: '${value}'\n`;
      } else {
        result += `${spaces}${key}: ${value}\n`;
      }
    } else if (typeof value === 'boolean') {
      result += `${spaces}${key}: ${value}\n`;
    } else if (typeof value === 'number') {
      result += `${spaces}${key}: ${value}\n`;
    } else if (Array.isArray(value)) {
      if (value.length === 0) {
        result += `${spaces}${key}: []\n`;
      } else if (typeof value[0] === 'object') {
        result += `${spaces}${key}:\n`;
        value.forEach(item => {
          result += `${spaces}  - ${stringifyYAML(item, indent + 2).slice(spaces.length + 2)}`;
        });
      } else {
        result += `${spaces}${key}:\n`;
        value.forEach(item => {
          result += `${spaces}  - ${item}\n`;
        });
      }
    } else if (typeof value === 'object') {
      result += `${spaces}${key}:\n`;
      result += stringifyYAML(value, indent + 1);
    }
  }

  return result;
}

// Test scripts
const getTestScript = (folder, filename) => {
  const name = filename.replace('.request.yaml', '');
  
  if (folder === 'Auth') {
    if (name === 'Login') {
      return `pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response has token", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("token");
  pm.expect(json.token).to.be.a("string");
  pm.environment.set("token", json.token);
  pm.environment.set("artistToken", json.token);
});

pm.test("Response has user details", function () {
  const json = pm.response.json();
  pm.expect(json.user).to.have.property("id");
  pm.expect(json.user).to.have.property("email");
  pm.expect(json.user).to.have.property("role");
});`;
    } else if (name === 'Register' || name === 'Register Dashboard') {
      return `pm.test("Status code is 200 or 201", function () {
  pm.response.to.have.status([200, 201]);
});

pm.test("Response has token", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("token");
  pm.expect(json.token).to.be.a("string");
  if (json.token) pm.environment.set("token", json.token);
});

pm.test("Response has user data", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("user");
  pm.expect(json.user).to.have.property("id");
  pm.expect(json.user).to.have.property("email");
});`;
    }
  }

  // Default test script based on method
  if (name.startsWith('Get')) {
    return `pm.test("Status code is 200", function () {
  pm.response.to.have.status(200);
});

pm.test("Response is valid JSON", function () {
  const json = pm.response.json();
  pm.expect(json).to.be.an("object");
});

pm.test("Response has data", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("data");
});`;
  } else {
    return `pm.test("Status code is 200 or 201", function () {
  pm.response.to.have.status([200, 201]);
});

pm.test("Response has success property", function () {
  const json = pm.response.json();
  pm.expect(json).to.have.property("success");
});`;
  }
};

function enhanceRequestFile(filepath) {
  try {
    let content = fs.readFileSync(filepath, 'utf8');
    const folder = path.basename(path.dirname(filepath));
    const filename = path.basename(filepath);

    const testCode = getTestScript(folder, filename);

    // Update the test script section
    const scriptPattern = /scripts:\s*\n([\s\S]*?)(?=\n[^\s]|\Z)/;
    const newScriptYAML = `scripts:
  - type: afterResponse
    language: text/javascript
    code: |-
${testCode.split('\n').map(line => `      ${line}`).join('\n')}`;

    if (scriptPattern.test(content)) {
      content = content.replace(scriptPattern, newScriptYAML + '\n');
    } else {
      // Append scripts if not present
      content += '\n' + newScriptYAML + '\n';
    }

    fs.writeFileSync(filepath, content, 'utf8');
    return { success: true, message: "Enhanced" };
  } catch (error) {
    return { success: false, message: error.message };
  }
}

function getAllRequestFiles(dir) {
  let files = [];
  const items = fs.readdirSync(dir);

  items.forEach(item => {
    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      files = files.concat(getAllRequestFiles(fullPath));
    } else if (item.endsWith('.request.yaml')) {
      files.push(fullPath);
    }
  });

  return files;
}

function main() {
  const basePath = path.join(__dirname, 'collections', 'EchoVault API');

  if (!fs.existsSync(basePath)) {
    console.log(`Collection path not found: ${basePath}`);
    return;
  }

  const requestFiles = getAllRequestFiles(basePath);

  console.log(`Found ${requestFiles.length} request files\n`);

  let successCount = 0;
  requestFiles.sort().forEach(filepath => {
    const folder = path.basename(path.dirname(filepath));
    const filename = path.basename(filepath);

    const result = enhanceRequestFile(filepath);

    const status = result.success ? "✓" : "✗";
    console.log(`${status} ${folder.padEnd(20)} | ${filename.padEnd(40)} | ${result.message}`);

    if (result.success) {
      successCount++;
    }
  });

  console.log(`\n${'='.repeat(80)}`);
  console.log(`✓ Enhanced ${successCount}/${requestFiles.length} request files`);
  console.log(`✓ Added/updated test scripts on all endpoints`);
}

main();
